File: python_cpp/uPIMulator_frontend/src/abi/isa/cc/sub_set_cc.py
from typing import Set

from abi.isa.cc._base_cc import BaseCC
from abi.isa.instruction.condition import Condition


class SubSetCC(BaseCC):
    def __init__(self, condition: Condition):
        super().__init__(
            SubSetCC.conditions(), condition,
        )

    @staticmethod
    def conditions() -> Set[Condition]:
        return {
            Condition.Z,
            Condition.NZ,
            Condition.XZ,
            Condition.XNZ,
            Condition.EQ,
            Condition.NEQ,
        }

File: golang_vm/uPIMulator/src/device/linker/kernel/instruction/cc/sub_set_cc.go
package cc

import (
	"errors"
)

type SubSetCc struct {
	condition Condition
}

func (this *SubSetCc) Init(condition Condition) {
	conditions := map[Condition]bool{
		Z:   true,
		NZ:  true,
		XZ:  true,
		XNZ: true,
		EQ:  true,
		NEQ: true,
	}

	if _, found := conditions[condition]; !found {
		err := errors.New("condition is not allowed")
		panic(err)
	}

	this.condition = condition
}

func (this *SubSetCc) Condition() Condition {
	return this.condition
}
